### ISSUE TITLE
questions about iteration order

### ISSUE THREAD

**Issue by hescluke** [Lai-YT/CS143-cool-compiler-assignments#1]
Excellent work! 
I got some questions about PA4. I notice that your ClassTable inherits from map. The entries in map are stored in the order of key, not the order of inserting. So when your semant walks through it in CheckFeatures, it should prints errors in the order of key instead of line numbers. But your semant works well and prints errors in the order of line numbers as same as the official semant. Can you tell me why it works? Thanks for your help!

**Comment by Lai-YT** [comment]
This is a great question!
I'm not 100% sure of the answer, but I would like to talk about my understanding of this issue.😄

As you mentioned, `std::map` sorts the elements (name-class pair) with their keys in ascending order, which are `Symbol`s in our case.
https://github.com/Lai-YT/CS143-cool-compiler-assignments/blob/ee565df5034c0b8a632f75c7122bf138bf8f36d1/assignments/PA4/semant.h#L29
A `Symbol` is, in fact, an `Entry *` (pointer!), which means the class (`Class_`) that has a lower **address** on its name (`Symbol`) will be touched first in an iteration. 

So, what's the relation between line numbers and addresses? To answer this, let's travel back to _PA2_ and _PA3_.😎 
In _PA2_, the lexical analysis phase, whenever we recognize a `TYPEID`, we add it to the `idtable`. Then, a **heap-allocated** `Symbol` is returned to us. 
https://github.com/Lai-YT/CS143-cool-compiler-assignments/blob/ee565df5034c0b8a632f75c7122bf138bf8f36d1/assignments/PA2/cool.flex#L169-L172
Such a `Symbol` later becomes the name of the `Class_` in _PA3_, the syntax analysis phase.
https://github.com/Lai-YT/CS143-cool-compiler-assignments/blob/ee565df5034c0b8a632f75c7122bf138bf8f36d1/assignments/PA3/cool.y#L201-L210
The key points are:
- **The `TYPEID` recognized first is added into the `idtable` first, and so is heap-allocated first**.
How can a `TYPEID` be recognized first? It has to appear earlier in the file, which indicates a smaller line number. 
- **[heap grows from low address to high address](https://courses.engr.illinois.edu/cs225/fa2022/resources/stack-heap/#general-memory-layout)**

To sum up, classes that appear earlier in the file have their names allocated first with lower addresses, leading to being sorted with a smaller key.

I hope you like the answer. Feel free to point out if I made any mistakes, or ask if you have other questions.😀 
> I tried to have the same error messages as the official compiler without knowing the specific orders beforehand. So, it was much about trial and error, and my code is a bit messy.😢

**Comment by hescluke** [comment]
So, you mean that what stored in the ClassTable is actually the address(pointer) of entries in the Class_ instead of their names. Since the order of address corresponds with the order of insertion, its iteration order is correct.

I don't know whether I understand your meaning properly. But if so, I think your explanation really makes sense!🤔

Thanks for your help! And I'm really looking forward to your PA5🤣

**Comment by Lai-YT** [comment]
Yes, the way that a `Class_` holds its name is to store the `Symbol` (`Entry*`) which points to its name (`Entry`).
I think it's a bit misleading to `typedef` an `Entry*` as `Symbol` since you might forget that it's a pointer.

I haven't scheduled _PA5_ yet since the memory leak issue in _PA4_ isn't solved (that's why _PA4_ is marked as _test passed_ instead of _done_), but I shall move on soon. 😁

**Comment by hescluke** [comment]
Anyway I really appreciate your help and good luck to your program.

**Comment by Lai-YT** [comment]
I'm happy that you like it. 🚀
